README.md
# Zeroconf JmDNS #

This repository packages android friendly sources from the [jmdns](http://jmdns.sourceforge.net/) project
and provides some tutorial demos.

See [rosjava_core](https://github.com/rosjava/rosjava_core) readme for
more information and pointers on the overall rosjava and android ecosystem.

## The JmDNS Library ##

### About ###

JmDNS works a little differently from avahi, although it does the same job. You can separate its
functionality into two parts:

1) Zeroconf services (publishing, discovery and resolution) on a fixed, specified interface.
2) Network topology aware zeroconf services (in the author's language - multi-homed).

What the latter means, is that it will be automatically detect what network interfaces you have
and also be aware of those interfaces going up and down whilst providing a dynamic awareness of
the zeroconf services that appear and disappear with those interfaces.

The network topology part of jmdns is still fairly experimental.

### The Experimental Part ###

The network topology aware part of zeroconf is quite important. On android in particular it saves
you having to worry about detecting the ip of your interface, which is rather handy. This part is
quite reliable after having helped with some patches upstream.

It does have problems however handling multiple network interfaces and can also be confusing with the
way it handles ipv6. I have some patches here in this code set which get it working, but either aren't
upstream yet, or probably not the proper solution yet.

Hence we have a snapshot of the sources here.

### Publishing ###

Publishing is relatively straightforward. For ros environments, we typically only ever want or need to
publish the ros master, in which case you'd publish it under something like:

```
_ros-master._tcp 11311
```

### Discovery ###

There are two methods in the jmdns library for discovery. The first is via *polling*, which will always be
the far more reliable method (especially if your interface is up and down alot and likely to get
jmdns confused). The second is via event based *listeners* which give you callbacks to instantly
react on a zeroconf service coming up or down.

## The ROS Wrapper ##

I've got a couple of classes set up to more easily handle the jmdns library for our purposes (jmdns
can be quite messy to program with). It's not a full node, but it wouldn't be a stretch to
convert this into a full node's functionality using zeroconf_msgs as the basis for handles that work
in a similar way to the ros zeroconf_avahi package.

## Tutorials ##

Tutorials are implemented in the jmdns_tutorials project. Instructions can be found
in [the zerconf_jmdns_suite package ROS wiki](http://wiki.ros.org/zeroconf_jmdns_suite).


CHANGELOG.rst
Changelog
=========

0.3.1 (2019-01-23)
------------------
* Updating Gradle version to 4.10.2.
* Fixing package name for jmdns package.
* Fix for tutorial links and testing instructions.
* Contributors: Daniel Stonier, Juan Ignacio Ubeira, Julian Cerruti

0.3.0 (2016-12-17)
------------------
* Updates for kinetic release.

0.2.1 (2015-02-25)
------------------
* gradle 1.8 -> 2.2.1
* Contributors: Daniel Stonier

0.2.0 (2015-02-18)
------------------
* First build on indigo

0.1.13 (2013-11-01)
-------------------
* constrain open ranged dependencies.

0.1.12 (2013-10-31)
-------------------
* use ROS_MAVEN_REPOSITORY

0.1.11 (2013-10-25)
-------------------
* official maven style open ended dependencies.

0.1.10 (2013-09-23)
-------------------
* use updated ros gradle plugins with maven-publish for publishing.

0.1.9 (2013-09-22)
------------------
* removing osgi information.

0.1.8 (2013-09-17)
------------------
* bugfix install rule target name.
* depend on rosjava_bootstrap for gradle plugins.
* gradle wrapper 1.7
* use catkin package xml version information
* gradled, maven'd, deb'd.
* ros-java -> rosjava migration.
* refactoring for unique java package names.
* fix badly copied package name for jdmns tutorial.
* catkinization using rosjava_tools
* demos working with new gradle build.
* cleared out non jmdns packages.

0.1.7 (2013-04-29)
------------------
* point at legacy rosinstallers.
* using multiproject gradle now so we can add more projects here.
* rosinstall file for rosjava.
* bugfix for name in package.
* readme added.
* now building and installing with catkin.
* first commit pulled in from kazuto's hg clone.
* avahi and comms moved out.

0.1.6 (2012-07-17)
------------------
* bugfix version bump.
* refactored for ros code style format.
* bugfix bad array size check, closes `#35 <https://github.com/rosjava/zeroconf_jmdns_suite/issues/35>`_.
* rosinstaller, use https, not git.
* updated fuerte rosinstaller.
* zeroconf: updated the rosinstaller for fuerte.

0.1.5 (2012-05-24)
------------------
* updated rosdeps for rosdep2.

0.1.4 (2012-04-12)
------------------
* Adjusting cmakelists for release into fuerte.
* hoping to avert compilation of jmdns. Probably should divert to a separate stack.
* new rosdeps.

0.1.3 (2012-01-04)
------------------
* jmdns : better getInstance call.
* jmdns: hack new jmmdns getInstance to always create, hack to trigger serviceRemoved callbacks on network down.
* jmdns_demos : forgot to add the publisher class.
* zeroconf_jmdns : added display names to the network interface logging.
* jmdns : put default listener before the listener instantiation (sometimes add services will come in before it gets set otherwise).
* zeroconf_jmdns : default listener to default discovery callback.
* zeroconf_jmdns : now hiding all service_info api.
* zeroconf_implementations : fix mediawiki formatting.
* zeroconf_avahi : fix code after update to zeroconf_comms ip4/ipv6 upgrade.
* zeroconf_jmdns: ros api utility toString function.
* zeroconf: split the ros api to include both ipv4 and ipv6, the jmdns is now working with that, but jmdns needs further updates to convert fully over to the ros comms representations.
* zeroconf_jmdns : now working rather more nicely, less multi interface confusion, but going to have to update everything else.
* this will break everything (new comms), but jmdns is looking better.
* zeroconf_jmdns_demos : better output.
* zeroconf_jmdns_demos : minor demo program refactoring.
* zeroconf_implementations : make sure we get resolving happening if a new service is found.
* zeroconf_jmdns_demos : master browser running, but strange resolving behaviour seen by jmdns.
* zeroconf_jmdns : added a default listener callback and api.
* zeroconf_jmdns : listeners with userland callback handlers.
* zeroconf_jmdns : user listener callbacks.
* zeroconf_jmdns : toString support for the demo.
* zeroconf_implementations : loggers so I can delete duplicating code in jmdns and android_jmdns
* zeroconf_jmdns : comment some debug spam.
* zeroconf_jmdns: simplify dependencies.
* zeroconf_implementations: forgot to update the build to the standard ros template.
* zeroconf_implementations : upgrade jmdns build method, fix android in rosinstalls.
* zeroconf_implementations : finalise the zeroconf_overlay.rosinstall.
* zeroconf: new overlay installer.

0.1.2 (2011-12-19)
------------------
* zeroconf: only maintain one rosinstaller, also bypass jmdns builds if rosjava_core 0.1.0 (old) is present.
* zeroconf_jmdns_demos : cleanup.
* zeroconf_implementations : decided to split the jmdns stack again (for illustration).





zeroconf_avahi/src/node/zeroconf.cpp
/**
 * @file /zeroconf_avahi/src/bin/zeroconf.cpp
 *
 * @brief File comment
 *
 * File comment
 *
 * @date 05/09/2011
 **/

/*****************************************************************************
 ** Includes
 *****************************************************************************/

#include <boost/function.hpp>

#include <ros/ros.h>
#include <zeroconf_comms/AddListener.h>
#include <zeroconf_comms/RemoveListener.h>
#include <zeroconf_comms/AddService.h>
#include <zeroconf_comms/ListDiscoveredServices.h>
#include <zeroconf_comms/ListPublishedServices.h>

#include "../../include/zeroconf_avahi/zeroconf.hpp"

/*****************************************************************************
 ** Namespaces
 *****************************************************************************/

using namespace zeroconf_avahi;
using namespace zeroconf_comms;

/*****************************************************************************
 ** Interface
 *****************************************************************************/

/**
 * @brief Wraps the library api into a fully functional ros node with ros api.
 *
 * Simply envelops the c++ api with ros msg/srv communications. Also has some
 * magic to help make publishing and browsing ros masters easy.
 */
class ZeroconfNode
{

private:
  typedef boost::function<void(zeroconf_comms::DiscoveredService)> connection_signal_cb;

public:
  void init(ros::NodeHandle &nh)
  {
    /*********************
     ** Ros Comms
     **********************/
    server_add_listener = nh.advertiseService("add_listener", &ZeroconfNode::add_listener, this);
    server_remove_listener = nh.advertiseService("remove_listener", &ZeroconfNode::remove_listener, this);
    server_add_service = nh.advertiseService("add_service", &ZeroconfNode::add_service, this);
    server_remove_service = nh.advertiseService("remove_service", &ZeroconfNode::remove_service, this);
    server_list_discovered_services = nh.advertiseService("list_discovered_services",
                                                          &ZeroconfNode::list_discovered_services, this);
    server_list_published_services = nh.advertiseService("list_published_services",
                                                         &ZeroconfNode::list_published_services, this);
    pub_new_connections = nh.advertise<DiscoveredService>("new_connections", 10);
    pub_lost_connections = nh.advertise<DiscoveredService>("lost_connections", 10);

    /*********************
     ** Signals
     **********************/
    // The callback functions use the publishers, so they must come after.
    connection_signal_cb new_connection_signal = std::bind1st(std::mem_fun(&ZeroconfNode::new_connections), this);
    connection_signal_cb old_connection_signal = std::bind1st(std::mem_fun(&ZeroconfNode::lost_connections), this);
    zeroconf.connect_signal_callbacks(new_connection_signal, old_connection_signal);

    /*********************
     ** Parameters
     **********************/
    ros::NodeHandle local_nh("~");
    // listeners
    XmlRpc::XmlRpcValue value;
    XmlRpc::XmlRpcValue::iterator iter;
    if (local_nh.getParam("listeners", value))
    {
      if (value.getType() != XmlRpc::XmlRpcValue::TypeArray)
      {
        ROS_ERROR_STREAM("Zeroconf: param variable 'listeners' has malformed type, should be of type array");
      }
      else
      {
        for (int i = 0; i < value.size(); ++i)
        {
          zeroconf.add_listener(value[i]);
        }
      }
    }
    // services
    XmlRpc::XmlRpcValue value_services;
    if (local_nh.getParam("services", value_services))
    {
      if (value_services.getType() != XmlRpc::XmlRpcValue::TypeArray)
      {
        ROS_ERROR_STREAM("Zeroconf: param variable 'services' has malformed type, should be of type array");
      }
      else
      {
        for (int i = 0; i < value_services.size(); ++i)
        {
          XmlRpc::XmlRpcValue value_service = value_services[i];
          if (value_service.getType() != XmlRpc::XmlRpcValue::TypeStruct)
          {
            ROS_ERROR_STREAM(
                "Zeroconf: " << i << "th element of param variable 'services' has malformed type, should be of type struct");
            break;
          }
          zeroconf_comms::PublishedService service;
          XmlRpc::XmlRpcValue value_name = value_service["name"];
          if (value_name.getType() != XmlRpc::XmlRpcValue::TypeString)
          {
            ROS_ERROR_STREAM("Zeroconf: services[" << i << "]['name'] has malformed type, should be of type string");
            break;
          }
          service.name = std::string(value_name);
          if (service.name == "Ros Master")
          {
            // add some special ros magic to easily identify master's.
            service.name = service.name + " (" + ros::master::getHost() + ")";
          }
          XmlRpc::XmlRpcValue value_type = value_service["type"];
          if (value_type.getType() != XmlRpc::XmlRpcValue::TypeString)
          {
            ROS_ERROR_STREAM("Zeroconf: services[" << i << "]['type'] has malformed type, should be of type string");
            break;
          }
          service.type = std::string(value_type);
          XmlRpc::XmlRpcValue value_domain = value_service["domain"];
          if (value_domain.getType() != XmlRpc::XmlRpcValue::TypeString)
          {
            ROS_ERROR_STREAM("Zeroconf: services[" << i << "]['domain'] has malformed type, should be of type string");
            break;
          }
          service.domain = std::string(value_domain);
          XmlRpc::XmlRpcValue value_port = value_service["port"];
          if (value_port.getType() == XmlRpc::XmlRpcValue::TypeInvalid)
          {
            service.port = ros::master::getPort();
            // wasn't set, so default to ros master's port
            ROS_WARN_STREAM(
                "Zeroconf: services[" << service.name << "]['port'] wasn't set, default to ros master's port [" << service.port << "]");
          }
          else if (value_port.getType() != XmlRpc::XmlRpcValue::TypeInt)
          {
            ROS_ERROR_STREAM(
                "Zeroconf: services[" << service.name << "]['port'] has malformed type, should be of type int");
            break;
          }
          else
          {
            service.port = value_port;
          }
          XmlRpc::XmlRpcValue value_description = value_service["description"];
          if (value_description.getType() != XmlRpc::XmlRpcValue::TypeString)
          {
            service.description = "";
          }
          else
          {
            service.description = std::string(value_description);
          }
          zeroconf.add_service(service);
        }
      }
    }
  }

private:
  bool add_listener(AddListener::Request &request, AddListener::Response &response)
  {
    ROS_DEBUG_STREAM("Zeroconf: serving an add_listener call... [" << request.service_type << "]");
    response.result = zeroconf.add_listener(request.service_type);
    return true;
  }
  bool remove_listener(RemoveListener::Request &request, RemoveListener::Response &response)
  {
    ROS_DEBUG_STREAM("Zeroconf: serving a remove_listener call... [" << request.service_type << "]");
    response.result = zeroconf.remove_listener(request.service_type);
    return true;
  }
  bool add_service(AddService::Request &request, AddService::Response &response)
  {
    response.result = zeroconf.add_service(request.service);
    return true;
  }
  bool remove_service(AddService::Request &request, AddService::Response &response)
  {
    response.result = zeroconf.remove_service(request.service);
    return true;
  }
  bool list_discovered_services(ListDiscoveredServices::Request &request, ListDiscoveredServices::Response &response)
  {
    ROS_DEBUG_STREAM("Zeroconf: serving a list_discovered_services call... [" << request.service_type << "]");
    zeroconf.list_discovered_services(request.service_type, response.services);
    return true;
  }
  bool list_published_services(ListPublishedServices::Request &request, ListPublishedServices::Response &response)
  {
    zeroconf.list_published_services(request.service_type, response.services);
    return true;
  }
  void new_connections(DiscoveredService service)
  {
    pub_new_connections.publish(service);
  }
  void lost_connections(DiscoveredService service)
  {
    pub_lost_connections.publish(service);
  }

  Zeroconf zeroconf;
  ros::ServiceServer server_add_listener, server_remove_listener, server_add_service, server_remove_service;
  ros::ServiceServer server_list_discovered_services, server_list_published_services;
  ros::Publisher pub_new_connections, pub_lost_connections;
};

/*****************************************************************************
 ** Main
 *****************************************************************************/

int main(int argc, char **argv)
{

  ros::init(argc, argv, "zeroconf");
  ros::NodeHandle nh;
  ZeroconfNode zeroconf_node;
  zeroconf_node.init(nh);
  ros::spin();
  return 0;
}
